src/Compiler/lex.fsl

```diff
@@ -1322,108 +1328,191 @@ and verbatimString sargs skip = parse
       else tripleQuoteString sargs skip lexbuf }
 
 // The rest is to break into pieces to allow double-click-on-word and other such things
  | ident
     { let (buf, _fin, m, kind, args) = sargs
       addUnicodeString buf (lexeme lexbuf)
-      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, m))
+      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, args.delimLength, m))
       else tripleQuoteString sargs skip lexbuf }
 
  | integer
  | xinteger
     { let (buf, _fin, m, kind, args) = sargs
       addUnicodeString buf (lexeme lexbuf)
-      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, m))
+      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, args.delimLength, m))
       else tripleQuoteString sargs skip lexbuf }
 
  | anywhite +
     { let (buf, _fin, m, kind, args) = sargs
       addUnicodeString buf (lexeme lexbuf)
-      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, m))
+      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, args.delimLength, m))
       else tripleQuoteString sargs skip lexbuf }
 
- | ("{{" | "}}")
+ | "%" +
     { let (buf, _fin, m, kind, args) = sargs
-      let s = lexeme lexbuf
-      addUnicodeString buf (if kind.IsInterpolated then s.[0..0] else s)
-      if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, m))
-      else tripleQuoteString sargs skip lexbuf }
+      let numPercents = lexeme lexbuf |> String.length
+      let result() =
+          if not skip then STRING_TEXT (LexCont.String(args.ifdefStack, args.stringNest, LexerStringStyle.TripleQuote, kind, args.delimLength, m))
+          else tripleQuoteString sargs skip lexbuf
+      if kind.IsInterpolated && args.delimLength > 1 then
+          // delimLength is number of $ chars prepended to opening quotes
+          // If number of consecutive % chars in content is equal to delimLength,
+          // then that sequence is treated as a format specifier,
+          // as in $"""%3d{42}""" or (equivalent) $$"""%%3d{{42}}""".
+          // Any extra % chars up to delimLength, are treated simply as regular string content.
+          // 2x delimLength or more % chars in a sequence will result in an error.
+          let maxPercents = 2 * args.delimLength - 1
+          if numPercents > maxPercents then
+              let m2 = lexbuf.LexemeRange
+              let rest = result()
+              args.diagnosticsLogger.ErrorR(Error(FSComp.SR.lexTooManyPercentsInTripleQuote(), m2))
```

Comment:
Well, the difference is going to be in the range of the error reported, basically
```
> $$$""" %%%%%%3d{{{42}}} """;;

  $$$""" %%%%%%3d{{{42}}} """;;
  -------^^^^^^
```
vs if `fail` is used:
```
> $$$""" %%%%%%3d{{{42}}} """;;

  $$$""" %%%%%%3d{{{42}}} """;;
  ^^^^^^^^^^^^^^^^^^
```
For many other cases we use range of a whole token, like in second example here, so that would be fine here too. I just saw an opportunity to make it more precise and took it.

I can be convinced to use `fail` instead.

Reply:
I guess the key is to be consistent with at least something - if we have this practice elsewhere, and it brings better precision, then let's keep it.

Reply:
I've just replaced a call to `args.diagnosticLogger.ErrorR` with `errorR`, which is used in couple different places in this file as well and produces the same effect.